posts/scala-for-the-arrogant-part-2.md

```diff
@@ -0,0 +1,427 @@
+
+# Refactoring Our Solution To Make It Scala
+Finally, we can rewrite this solution to use Scala. We will still retain the structure that we devised earlier, but let's see if the Scala syntax helps us create something more elegant. The inspiration for this solution comes from Martin Odersky's *[Functional Program Design In Scala](https://www.coursera.org/learn/progfun2)*, with some modifications made for clarity and to illustrate some different concepts.
+
+## One File
+Scala has a lot of nice features for development like worksheets: an interactive mode where you can see the results of your coding in real-time. We can also define a lot of our landscape in a single file. We will take that approach here. It may not be best practice but it will allow us to keep it simple and hold our solution in one space.
+
+## Define our own "Primitives"
+In our Java solution we continuously referred to our state by a list of integers (with each integer representing the volume of water currently in a jug) and our jugs as a list of integers (with each integer representing the capacity of the jug). Passing around two lists of ints with different meanings could be confusing and cause error. One alternative is to define new types that shadow the actual data underneath:
+```scala
+object WaterPouringTypes {
+  type Jug = Int
+  type State = List[Int]
+  type ActionTrail = List[List[Action]]
+}
+```
+Now we can pass a `Jug` as a parameter, have functions explicitly request a jug instead of an int, still have our logic treat it as an Int. For example, we'll take the Move Interface from our Java solution and model it as a trait, using the `Jug` and `State` types we just defined:
+
+```scala
+trait Move {
+    def doMove(state: State): State
+}
+case class Empty(jugIndex: Jug) extends Move {
+    override def doMove(state: State): State = state updated (jugIndex, 0)
+}
+case class Fill(jugIndex: Jug) extends Move {
+    override def doMove(state: State): State = state updated (jugIndex, JUGS(jugIndex))
+}
+case class Pour(fromJugIndex: Jug, toJugIndex: Jug) extends Move {
+    override def doMove(state: State): State = {
+        val pourAmt = math.min(state(fromJugIndex), JUGS(toJugIndex) - state(toJugIndex))
+        state updated (fromJugIndex, state(fromJugIndex) - pourAmt) updated (toJugIndex, state(toJugIndex) + pourAmt)
+    }
+}  
+```
+You can see how concisely we are able to declare the trait and its implementations in the form of a case class. Using the `updated` method we can create a modified clone of state without modifying the input state. All of Scalas collections and datatypes are immutable by default so this is handled for us elegantly.
```

Comment:
"All of Scalas collections and datatypes are immutable by default so this is handled for us elegantly." -> Not really, there are 2 sets of collections: one for mutable collections, one for immutable ones.

Reply:
Yes, but the default is immutable, right? AFAIK it will not be mutable unless you explicitly state otherwise. I think I see what you are saying tho: the way I worded it could be confusing

Reply:
rewrote to make it clearer